File: frontend/src/lib/utils/lawyer.ts
import type { Tables } from '@/types/supabase'

type LawyerProfile = Tables<'lawyer_profiles'>

// ── Fee Range ─────────────────────────────────────────────
export function formatLawyerFeeRange(
    feeMin: number | null,
    feeMax: number | null
): string {
    if (!feeMin && !feeMax) return 'Fee on consultation'
    if (feeMin && !feeMax) return `From ₹${feeMin.toLocaleString('en-IN')}`
    if (!feeMin && feeMax) return `Up to ₹${feeMax.toLocaleString('en-IN')}`
    return `₹${feeMin!.toLocaleString('en-IN')} – ₹${feeMax!.toLocaleString('en-IN')}`
}

// ── Experience Label ──────────────────────────────────────
export function mapExperienceToLabel(years: number | null): string {
    if (!years) return 'Junior Advocate'
    if (years <= 2) return 'Junior Advocate (0–2 years)'
    if (years <= 5) return 'Associate Advocate (3–5 years)'
    if (years <= 10) return 'Mid-Level Advocate (6–10 years)'
    if (years <= 20) return 'Senior Advocate (11–20 years)'
    return 'Principal Advocate (20+ years)'
}

// ── Primary Specialisation ────────────────────────────────
export function getPrimarySpecialisation(
    specialisations: string[] | null
): string {
    if (!specialisations || specialisations.length === 0) return 'General Practice'
    return formatDomain(specialisations[0])
}

// ── Verification Flag ─────────────────────────────────────
export function isVerified(
    status: string | null
): boolean {
    return status === 'verified'
}

// ── Response Time Label ───────────────────────────────────
export function formatResponseTime(hours: number | null): string {
    if (!hours) return 'Responds within 24 hours'
    if (hours < 1) return 'Responds within 1 hour'
    if (hours === 1) return 'Responds within 1 hour'
    if (hours <= 24) return `Responds within ${hours} hours`
    const days = Math.ceil(hours / 24)
    return `Responds within ${days} day${days > 1 ? 's' : ''}`
}

// ── Domain Label Formatter ────────────────────────────────
export function formatDomain(domain: string): string {
    const map: Record<string, string> = {
        consumer: 'Consumer Law',
        tenant: 'Tenant / Rent',
        labour: 'Labour & Employment',
        criminal: 'Criminal Law',
        cyber: 'Cyber Law',
        property: 'Property Law',
        family: 'Family Law',
        rti: 'RTI',
        corruption: 'Anti-Corruption',
        civil: 'Civil Law',
        tax: 'Tax Law',
        corporate: 'Corporate / Business',
        intellectual_property: 'Intellectual Property',
        constitutional: 'Constitutional / PIL',
        other: 'General Practice'
    }
    return map[domain] ?? domain
}

// ── Map full LawyerProfile to UI shape ───────────────────
export function mapLawyerToCard(lawyer: LawyerProfile) {
    return {
        id: lawyer.id,
        name: lawyer.full_name ?? 'Unknown',
        title: lawyer.professional_title ?? getPrimarySpecialisation(lawyer.specialisations),
        primaryCategory: lawyer.primary_category ?? getPrimarySpecialisation(lawyer.specialisations),
        specialisations: lawyer.specialisations ?? [],
        specialty: getPrimarySpecialisation(lawyer.specialisations),
        priceRange: formatLawyerFeeRange(lawyer.fee_min, lawyer.fee_max),
        experienceLabel: mapExperienceToLabel(lawyer.experience_years),
        experienceYears: lawyer.experience_years ?? 0,
        verified: isVerified(lawyer.verification_status),
        rating: lawyer.avg_rating ?? 0,
        responseTime: formatResponseTime(lawyer.response_time_hours),
        photoUrl: lawyer.profile_photo_url,
        state: lawyer.practice_state,
        district: lawyer.practice_district,
        languages: lawyer.languages ?? ['en'],
        feeMin: lawyer.fee_min,
        feeMax: lawyer.fee_max,
        isAvailable: lawyer.is_available,
    }
}

export type LawyerCard = ReturnType<typeof mapLawyerToCard>

File: backend/app/schemas/lawyer.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class LegalDomain(str, Enum):
    consumer              = "consumer"
    tenant                = "tenant"
    labour                = "labour"
    criminal              = "criminal"
    cyber                 = "cyber"
    property              = "property"
    family                = "family"
    rti                   = "rti"
    corruption            = "corruption"
    civil                 = "civil"
    tax                   = "tax"
    corporate             = "corporate"
    intellectual_property = "intellectual_property"
    constitutional        = "constitutional"
    other                 = "other"


class VerificationStatus(str, Enum):
    unverified = "unverified"
    pending    = "pending"
    verified   = "verified"
    rejected   = "rejected"


class LawyerProfileResponse(BaseModel):
    id:                   str
    full_name:            Optional[str]
    email:                Optional[str]
    phone:                Optional[str]
    professional_title:   Optional[str]
    primary_category:     Optional[str]
    bar_council_id:       Optional[str]
    practice_state:       Optional[str]
    practice_district:    Optional[str]
    court_types:          Optional[list[str]]
    specialisations:      Optional[list[LegalDomain]]
    experience_years:     Optional[int]
    bio:                  Optional[str]
    languages:            Optional[list[str]]
    fee_min:              Optional[int]
    fee_max:              Optional[int]
    response_time_hours:  Optional[int]
    profile_photo_url:    Optional[str]
    verification_status:  Optional[VerificationStatus]
    completeness_score:   int
    is_active:            bool
    is_available:         bool
    avg_rating:           Optional[float]
    total_cases:          Optional[int]
    win_rate:             Optional[float]
    total_reviews:        Optional[int]
    created_at:           str
    updated_at:           Optional[str]

    # Computed fields (not in DB — added by API)
    verified:             bool = Field(default=False)
    fee_range:            Optional[str] = None
    experience_label:     Optional[str] = None
    response_time_label:  Optional[str] = None

    def model_post_init(self, __context):
        self.verified = self.verification_status == VerificationStatus.verified

        if self.fee_min and self.fee_max:
            self.fee_range = f"₹{self.fee_min:,} – ₹{self.fee_max:,}"

        years = self.experience_years or 0
        if years <= 2:   self.experience_label = "Junior Advocate"
        elif years <= 5: self.experience_label = "Associate Advocate (3–5 years)"
        elif years <= 10:self.experience_label = "Mid-Level Advocate (6–10 years)"
        elif years <= 20:self.experience_label = "Senior Advocate (11–20 years)"
        else:            self.experience_label = "Principal Advocate (20+ years)"

        hours = self.response_time_hours or 24
        if hours <= 1:   self.response_time_label = "Responds within 1 hour"
        elif hours <= 24:self.response_time_label = f"Responds within {hours} hours"
        else:            self.response_time_label = f"Responds within {hours // 24} days"


class LawyerProfileUpdate(BaseModel):
    full_name:           Optional[str] = None
    professional_title:  Optional[str] = None
    primary_category:    Optional[str] = None
    phone:               Optional[str] = None
    bar_council_id:      Optional[str] = None
    practice_state:      Optional[str] = None
    practice_district:   Optional[str] = None
    court_types:         Optional[list[str]] = None
    specialisations:     Optional[list[LegalDomain]] = None
    experience_years:    Optional[int] = None
    bio:                 Optional[str] = None
    languages:           Optional[list[str]] = None
    fee_min:             Optional[int] = None
    fee_max:             Optional[int] = None
    response_time_hours: Optional[int] = None
    is_available:        Optional[bool] = None